test/core/bad_connection/close_fd_test.cc

```diff
@@ -0,0 +1,731 @@
+  op = ops;
+  op->op = GRPC_OP_SEND_INITIAL_METADATA;
+  op->data.send_initial_metadata.count = 0;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_SEND_MESSAGE;
+  op->data.send_message.send_message = request_payload;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_SEND_CLOSE_FROM_CLIENT;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_RECV_INITIAL_METADATA;
+  op->data.recv_initial_metadata.recv_initial_metadata = &initial_metadata_recv;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_RECV_MESSAGE;
+  op->data.recv_message.recv_message = &response_payload_recv;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_RECV_STATUS_ON_CLIENT;
+  op->data.recv_status_on_client.trailing_metadata = &trailing_metadata_recv;
+  op->data.recv_status_on_client.status = &status;
+  op->data.recv_status_on_client.status_details = &details;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  error = grpc_call_start_batch(c, ops, static_cast<size_t>(op - ops), tag(1),
+                                nullptr);
+  GPR_ASSERT(GRPC_CALL_OK == error);
+
+  error = grpc_server_request_call(ctx.server, &s, &call_details,
+                                   &request_metadata_recv, ctx.bound_cq, ctx.cq,
+                                   tag(101));
+  GPR_ASSERT(GRPC_CALL_OK == error);
+  e = grpc_completion_queue_next(
+      ctx.cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  GPR_ASSERT(e.success == 1);
+  GPR_ASSERT(e.tag == tag(101));
+  GPR_ASSERT(e.type == GRPC_OP_COMPLETE);
+
+  memset(ops, 0, sizeof(ops));
+  op = ops;
+  op->op = GRPC_OP_SEND_INITIAL_METADATA;
+  op->data.send_initial_metadata.count = 0;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_RECV_MESSAGE;
+  op->data.recv_message.recv_message = &request_payload_recv;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+
+  grpc_endpoint_pair* sfd = ctx.ep;
+  int fd;
+  if (f == SERVER_FD) {
+    fd = sfd->server->vtable->get_fd(sfd->server);
+  } else {
+    GPR_ASSERT(f == CLIENT_FD);
+    fd = sfd->client->vtable->get_fd(sfd->client);
+  }
+  /* Connection is closed before the server receives the client's message. */
+  close(fd);
+
+  error = grpc_call_start_batch(s, ops, static_cast<size_t>(op - ops), tag(102),
+                                nullptr);
+  GPR_ASSERT(GRPC_CALL_OK == error);
+
+  /* Batch operation succeeds on the server side */
+  e = grpc_completion_queue_next(
+      ctx.bound_cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  GPR_ASSERT(e.type == GRPC_OP_COMPLETE);
+  GPR_ASSERT(e.success == 1);
+  GPR_ASSERT(e.tag == tag(102));
+
+  e = grpc_completion_queue_next(
+      ctx.client_cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  if (f == CLIENT_FD) {
+    /* Timeout is returned to the client whe the client (local) fd is closed */
+    GPR_ASSERT(e.type == GRPC_QUEUE_TIMEOUT);
```

Comment:
Do we know if it is the socket not emitting any event or it is something else that caused TIMEOUT in this case?

Reply:
Looks like `GRPC_QUEUE_TIMEOUT` in  a couple of cases:
1. Deadline exceeded, i.e. socket didn't emit any event. https://github.com/grpc/grpc/blob/master/src/core/lib/surface/completion_queue.cc#L1031
2. Completion queue next failed. I'm not sure when that condition will be hit. https://github.com/grpc/grpc/blob/master/src/core/lib/surface/completion_queue.cc#L1049